File: polyglot/csharp/layer_size_calculator.cs
using System;
using System.Collections.Generic;
from System.IO import Path, File, StreamReader;
using System.Text.RegularExpressions;
using System.Linq;

namespace shipcheck.polyglot.csharp;

/// <summary>
/// Calculates expected layer sizes for a Dockerfile before building.
/// Uses empirical compression ratios and instruction heuristics.
/// </summary>
public static class LayerSizeCalculator
{
    private const double TextCompressionRatio = 0.35; // ~65% reduction after gzip
    private const double BinaryCompressionRatio = 0.70; // ~30% reduction
    private const int DefaultLayerOverhead = 1024; // Metadata overhead per layer

    public record LayerInfo(
        string Instruction,
        long CompressedSize,
        long CumulativeSize,
        string? Notes
    );

    /// <summary>
    /// Main entry point for the calculator.
    /// </summary>
    public static void RunDemo()
    {
        var dockerfileContent = @"FROM node:18-alpine AS base
WORKDIR /app
COPY package*.json ./
RUN npm ci --only=production
COPY . .
EXPOSE 3000
CMD [""node"", ""dist/index.js""]";

        var layers = CalculateLayerSizes(dockerfileContent);
        
        Console.WriteLine("=== Layer Size Analysis ===");
        PrintLayerSummary(layers);
    }

    /// <summary>
    /// Parses a Dockerfile and returns per-layer size estimates.
    /// </summary>
    public static List<LayerInfo> CalculateLayerSizes(string dockerfileContent)
    {
        var lines = dockerfileContent.Split(new[] { "\r\n", "\n" }, StringSplitOptions.None);
        var layers = new List<LayerInfo>();
        
        long cumulativeSize = 0;
        string? currentInstruction = null;
        bool inMultiLineCommand = false;

        for (int i = 0; i < lines.Length; i++)
        {
            var line = lines[i].Trim();

            // Skip empty lines and comments
            if (string.IsNullOrEmpty(line) || line.StartsWith("#"))
                continue;

            // Handle multi-line commands (backslash continuation)
            if (inMultiLineCommand)
            {
                currentInstruction += " " + line.Trim();
                inMultiLineCommand = !line.EndsWith("\\");
                continue;
            }

            // Check for line continuation
            if (line.EndsWith("\\"))
            {
                inMultiLineCommand = true;
                currentInstruction = line.Remove(line.Length - 1).Trim();
                continue;
            }

            // Parse instruction and arguments
            var parts = ExtractInstructionAndArgs(line);
            var instruction = parts.Key;
            var args = parts.Value;

            if (string.IsNullOrEmpty(instruction))
                continue;

            // Calculate layer size based on instruction type
            long estimatedSize;
            string? notes = null;

            switch (instruction.ToUpperInvariant())
            {
                case "FROM":
                    // FROM creates a new base image - use default overhead
                    estimatedSize = DefaultLayerOverhead;
                    notes = $"Base image: {args}";
                    break;

                case "WORKDIR":
                case "ENV":
                case "LABEL":
                case "EXPOSE":
                case "ARG":
                    // Small metadata changes
                    estimatedSize = 512 + CalculateStringSize(args);
                    notes = $"Metadata: {instruction}";
                    break;

                case "RUN":
                    // Command output varies - use heuristic based on command length
                    var cmdLength = args.Length;
                    estimatedSize = DefaultLayerOverhead + 
                                   (cmdLength * 256) + // Rough estimate per word
                                   CalculateStringSize(args);
                    
                    notes = $"Command: {args}";
                    
                    // Heuristic for common commands
                    if (args.Contains("npm ci") || args.Contains("yarn install"))
                        notes += " | npm/yarn cache included";
                    else if (args.Contains("apt-get") || args.Contains("apk add"))
                        notes += " | package manager output included";
                    
                    break;

                case "COPY":
                case "ADD":
                    // Estimate based on source files specified
                    var sources = ExtractSources(args);
                    estimatedSize = DefaultLayerOverhead + 
                                   CalculateStringSize(string.Join(" ", sources));
                    
                    notes = $"Copy: {sources.Length} items";
                    
                    // Heuristic for common patterns
                    if (sources.Any(s => s.Contains(".json") || s.Contains(".js")))
                        notes += " | likely source code";
                    else if (sources.Any(s => s.Contains("node_modules")))
                        notes += " | node_modules included";
                    
                    break;

                case "CMD":
                case "ENTRYPOINT":
                    estimatedSize = 512 + CalculateStringSize(args);
                    notes = $"Entry point: {instruction}";
                    break;

                default:
                    // Unknown instruction - conservative estimate
                    estimatedSize = DefaultLayerOverhead + 
                                   CalculateStringSize(instruction) + 
                                   CalculateStringSize(args);
                    notes = $"Unknown: {instruction}";
                    break;
            }

            cumulativeSize += estimatedSize;

            layers.Add(new LayerInfo(
                instruction,
                estimatedSize,
                cumulativeSize,
                notes
            ));
        }

        return layers;
    }

    /// <summary>
    /// Extracts the main instruction and its arguments.
    /// </summary>
    private static (string Instruction, string Args) ExtractInstructionAndArgs(string line)
    {
        // Handle inline comments
        var commentIndex = line.IndexOf("//");
        if (commentIndex >= 0)
            line = line.Substring(0, commentIndex).Trim();

        // Find first whitespace after instruction name
        var spaceIndex = line.IndexOf(' ');
        string instruction;
        string args;

        if (spaceIndex > -1)
        {
            instruction = line.Substring(0, spaceIndex);
            args = line.Substring(spaceIndex).Trim();
        }
        else
        {
            instruction = line;
            args = "";
        }

        return (instruction, args);
    }

    /// <summary>
    /// Extracts source file paths from COPY/ADD arguments.
    /// </summary>
    private static List<string> ExtractSources(string args)
    {
        var sources = new List<string>();
        
        // Simple heuristic: split by whitespace, filter out flags
        var tokens = args.Split(new[] { " ", "\t" }, StringSplitOptions.RemoveEmptyEntries);
        
        foreach (var token in tokens)
        {
            if (!IsFlag(token))
                sources.Add(token);
        }

        return sources;
    }

    private static bool IsFlag(string token)
    {
        // Common COPY/ADD flags that indicate options rather than files
        var flagPatterns = new[] 
        {
            "-t", "--from=", "--chown=", "--chmod=",
            "/app/", "/", "./", "../"
        };

        return flagPatterns.Any(p => token.StartsWith(p, StringComparison.OrdinalIgnoreCase));
    }

    /// <summary>
    /// Estimates string size based on character count and typical encoding.
    /// </summary>
    private static long CalculateStringSize(string input)
    {
        if (string.IsNullOrEmpty(input))
            return 0;

        // Estimate: each character ~1 byte, plus some overhead for escaping/encoding
        var charCount = input.Length;
        var estimatedBytes = charCount + (charCount / 4); // Extra for escapes
        
        return Math.Max(estimatedBytes, 64); // Minimum size
    }

    /// <summary>
    /// Prints a summary of all calculated layers.
    /// </summary>
    private static void PrintLayerSummary(List<LayerInfo> layers)
    {
        Console.WriteLine();
        
        var totalSize = layers.Sum(l => l.CompressedSize);
        var maxLayer = layers.OrderByDescending(l => l.CompressedSize).First();

        Console.WriteLine($"Total estimated size: {(totalSize / 1024):F1} KB");
        Console.WriteLine($"Largest layer: {maxLayer.Instruction} ({(maxLayer.CompressedSize / 1024):F1} KB)");
        
        if (layers.Count > 0)
        {
            var avgSize = layers.Average(l => l.CompressedSize);
            Console.WriteLine($"Average layer size: {(avgSize / 1024):F2} KB");
        }

        Console.WriteLine();
        Console.WriteLine("Layer breakdown:");
        
        foreach (var layer in layers)
        {
            var bar = GetProgressBar(layer.CompressedSize, totalSize);
            Console.WriteLine($"  [{bar}] {layer.Instruction,-15} {(layer.CompressedSize / 1024):F1} KB");
            
            if (!string.IsNullOrEmpty(layer.Notes))
                Console.WriteLine($"      Note: {layer.Notes}");
        }

        // Add recommendations
        PrintRecommendations(layers, totalSize);
    }

    private static string GetProgressBar(long current, long max)
    {
        var width = 40;
        var ratio = (double)Math.Min(current, max) / Math.Max(1, max);
        var filled = (int)(width * ratio);
        
        return new string('#', filled) + new string('-', width - filled);
    }

    private static void PrintRecommendations(List<LayerInfo> layers, long totalSize)
    {
        Console.WriteLine();
        Console.WriteLine("=== Recommendations ===");

        var recommendations = new List<string>();

        // Check for large COPY operations
        if (layers.Any(l => l.Instruction == "COPY" && l.CompressedSize > 1024 * 5))
            recommendations.Add("- Consider using multi-stage builds to reduce final image size.");

        // Check for node_modules in COPY
        var hasNodeModules = layers.Any(l => 
            l.Instruction == "COPY" && l.Notes?.Contains("node_modules") == true);
        
        if (hasNodeModules)
            recommendations.Add("- Found node_modules being copied. Ensure 'npm ci' runs before COPY.");

        // Check for large base image
        var hasLargeBase = layers.Any(l => 
            l.Instruction == "FROM" && l.CompressedSize > 1024 * 5);
        
        if (hasLargeBase)
            recommendations.Add("- Base image is relatively large. Consider using distroless or slim variants.");

        // Check for many RUN commands
        var runCount = layers.Count(l => l.Instruction == "RUN");
        if (runCount > 3)
            recommendations.Add($"- Found {runCount} RUN instructions. Consolidate where possible to reduce layer count.");

        // General advice
        if (totalSize > 1024 * 50)
            recommendations.Add("- Total size exceeds 50KB. Consider multi-stage builds for production.");

        if (!recommendations.Any())
            recommendations.Add("Image appears well-optimized based on heuristics.");

        foreach (var rec in recommendations)
            Console.WriteLine($"  {rec}");
    }
}

// Entry point for running the demo
public class Program
{
    public static void Main(string[] args)
    {
        LayerSizeCalculator.RunDemo();
    }
}

File: polyglot/c/layer_size_calculator.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <sys/stat.h>

#define MAX_LINE 4096
#define MAX_LAYERS 1024
#define BASE_IMAGE_SIZE (50 * 1024 * 1024) /* 50MB default */

typedef struct {
    char name[256];
    size_t estimated_size;
    int is_base;
} LayerInfo;

static LayerInfo layers[MAX_LAYERS];
static int layer_count = 0;
static char current_layer_name[256] = "";
static long long total_estimated_size = BASE_IMAGE_SIZE;

/* Get file size in bytes */
static size_t get_file_size(const char *path) {
    struct stat st;
    if (stat(path, &st) == 0) {
        return (size_t)st.st_size;
    }
    return 1024; /* Default for unknown files */
}

/* Parse a single Dockerfile line and update layer tracking */
static void parse_line(const char *line) {
    int i, j;
    char cmd[256] = "";
    char args[MAX_LINE];
    
    if (strlen(line) == 0 || line[0] == '#') return;

    /* Trim leading whitespace and get command */
    for (i = 0; i < MAX_LINE && isspace((unsigned char)line[i]); i++) {
        ;
    }
    if (line[i] == '\0' || line[i] == '#') return;

    strncpy(cmd, &line[i], sizeof(cmd) - 1);
    cmd[sizeof(cmd) - 1] = '\0';

    /* Find end of command */
    for (i = 0; i < MAX_LINE && !isspace((unsigned char)line[i]); i++) {
        ;
    }
    strncpy(args, &line[i], sizeof(args) - 1);
    args[sizeof(args) - 1] = '\0';

    /* Handle FROM instruction */
    if (strncmp(cmd, "FROM", 4) == 0) {
        char image[256];
        int found_space = 0;
        
        for (i = 0; i < sizeof(image); i++) {
            if (!found_space && isspace((unsigned char)cmd[i])) {
                break;
            }
            image[i] = cmd[i];
            if (cmd[i] == ' ') found_space = 1;
        }
        image[found_space ? sizeof(image) - 1 : i] = '\0';

        /* Reset layer tracking for new base */
        if (layer_count > 0) {
            layers[layer_count - 1].estimated_size += total_estimated_size;
        }
        
        strncpy(current_layer_name, image, sizeof(current_layer_name) - 1);
        current_layer_name[sizeof(current_layer_name) - 1] = '\0';
        layer_count = 0;
        total_estimated_size = BASE_IMAGE_SIZE;

        printf("  [BASE] %s\n", image);
        return;
    }

    /* Handle ADD and COPY */
    if (strncmp(cmd, "ADD ", 4) == 0 || strncmp(cmd, "COPY ", 5) == 0) {
        char src[256];
        long long file_size = 1024;
        
        /* Extract source path */
        for (i = 0; i < MAX_LINE && !isspace((unsigned char)args[i]); i++) {
            ;
        }
        strncpy(src, args, sizeof(src) - 1);
        src[sizeof(src) - 1] = '\0';

        /* Check if source is a file in current directory */
        size_t fs = get_file_size(src);
        
        printf("  [COPY/ADD] %s -> +%.2f MB\n", 
               src, (double)fs / 1048576.0);
        
        total_estimated_size += fs;
        layer_count++;

        /* Track this layer */
        strncpy(layers[layer_count - 1].name, current_layer_name, sizeof(layers[0].name) - 1);
        layers[layer_count - 1].estimated_size = total_estimated_size;
        layers[layer_count - 1].is_base = (layer_count == 1);

        return;
    }

    /* Handle ENV and ARG */
    if (strncmp(cmd, "ENV ", 4) == 0 || strncmp(cmd, "ARG ", 4) == 0) {
        printf("  [META] %s\n", cmd);
        return;
    }

    /* Handle RUN - estimate based on command length */
    if (strncmp(cmd, "RUN ", 4) == 0) {
        long long run_estimate = 256 * 1024; /* Default 256KB for shell commands */
        
        /* Heuristic: longer commands might produce more output */
        size_t cmd_len = strlen(args);
        if (cmd_len > 100) {
            run_estimate += (cmd_len - 100) * 4096;
        }

        printf("  [RUN] %s\n", args);
        printf("    Est. output: +%.2f MB\n", (double)run_estimate / 1048576.0);
        
        total_estimated_size += run_estimate;
        layer_count++;

        strncpy(layers[layer_count - 1].name, current_layer_name, sizeof(layers[0].name) - 1);
        layers[layer_count - 1].estimated_size = total_estimated_size;
        layers[layer_count - 1].is_base = (layer_count == 1);

        return;
    }

    /* Handle other instructions */
    if (strncmp(cmd, "RUN ", 4) != 0 && 
        strncmp(cmd, "COPY ", 5) != 0 && 
        strncmp(cmd, "ADD ", 4) != 0 &&
        strncmp(cmd, "FROM", 4) != 0 &&
        strncmp(cmd, "ENV ", 4) != 0 &&
        strncmp(cmd, "ARG ", 4) != 0) {
        
        /* Default estimate for unknown instructions */
        printf("  [OTHER] %s\n", cmd);
    }
}

/* Print final summary */
static void print_summary(void) {
    printf("\n=== LAYER SIZE SUMMARY ===\n");
    printf("Total estimated size: %.2f MB\n\n", (double)total_estimated_size / 1048576.0);
    
    if (layer_count > 0) {
        printf("Layer breakdown:\n");
        for (int i = 0; i < layer_count; i++) {
            printf("  %2d: %.2f MB\n", 
                   i + 1,
                   (double)(layers[i].estimated_size - layers[i > 0 ? i - 1 : 0].is_base ? 0 : 0) / 1048576.0);
        }
    }
}

/* Parse entire Dockerfile */
static int parse_dockerfile(const char *filename) {
    FILE *fp;
    char line[MAX_LINE];
    int result = 0;

    fp = fopen(filename, "r");
    if (!fp) {
        fprintf(stderr, "Error: Cannot open file '%s'\n", filename);
        return 1;
    }

    printf("Parsing Dockerfile: %s\n", filename);
    printf("\n--- LAYER ANALYSIS ---\n");

    while (fgets(line, sizeof(line), fp)) {
        parse_line(line);
    }

    fclose(fp);
    print_summary();
    
    return result;
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Usage: %s <Dockerfile>\n", argv[0]);
        printf("\nExample:\n");
        printf("  %s Dockerfile\n", argv[0]);
        
        /* Demo with embedded Dockerfile */
        printf("\n--- DEMO MODE (embedded Dockerfile) ---\n");
        
        const char *demo_dockerfile = 
            "FROM ubuntu:22.04\n"
            "ENV APP=shipcheck\n"
            "RUN apt-get update && apt-get install -y curl wget\n"
            "COPY . /app/\n"
            "ADD https://example.com/config.yaml /etc/app/\n"
            "RUN make build\n"
            "CMD [\"./app\"]\n";

        /* Write demo to temp file */
        FILE *tmp = tmpfile();
        fprintf(tmp, "%s", demo_dockerfile);
        rewind(tmp);
        
        char buffer[4096];
        size_t n;
        while ((n = fread(buffer, 1, sizeof(buffer), tmp)) > 0) {
            parse_line((char*)buffer);
        }
        fclose(tmp);

        printf("\nDemo complete.\n");
    } else {
        parse_dockerfile(argv[1]);
    }

    return 0;
}